File: packages/sdk/examples/06-retry-and-cancel.ts
/**
 * 06-retry-and-cancel.ts — Retry policy + AbortSignal threading.
 *
 * Demonstrates: custom retry policy, onRetry telemetry hook,
 * AbortSignal cancellation, distinct error classes
 * (TextralRetryExhausted vs the underlying error).
 *
 * Env: same as 01-quick-start.ts.
 *
 * Run:
 *     npx tsx 06-retry-and-cancel.ts
 *
 * See also:
 *     SDK_COOKBOOK_OUTLINE.md §7
 */

import {
  TextralClient,
  TextralRetryExhausted,
  TextralApiError,
} from '@textral/sdk';

const profileName = process.env.TEXTRAL_PROFILE;
const namespace = process.env.TEXTRAL_NAMESPACE ?? 'cookbook';
const providerKeyRef = process.env.TEXTRAL_PROVIDER_KEY_REF ?? 'openai';

const client = profileName
  ? new TextralClient({
      profile: profileName,
      retry: {
        maxAttempts: 5,
        initialDelayMs: 500,
        maxDelayMs: 8_000,
        retryOn: [429, 502, 503, 504],
        onRetry: ({ attempt, status, delayMs }) => {
          console.error(
            `retry #${attempt}: status=${status ?? 'transport'} sleeping ${delayMs}ms`,
          );
        },
      },
    })
  : new TextralClient({
      baseUrl: process.env.TEXTRAL_BASE_URL ?? 'https://api.textral.alacrity.ai',
      apiKey: process.env.TEXTRAL_API_KEY ?? '',
      retry: {
        maxAttempts: 5,
        initialDelayMs: 500,
        maxDelayMs: 8_000,
        retryOn: [429, 502, 503, 504],
        onRetry: ({ attempt, status, delayMs }) => {
          console.error(
            `retry #${attempt}: status=${status ?? 'transport'} sleeping ${delayMs}ms`,
          );
        },
      },
    });

const ac = new AbortController();
// Cancel after 5s — beats anything but a fast-path success.
setTimeout(() => ac.abort(), 5_000).unref();

try {
  const r = await client.query(
    {
      namespace,
      query: 'What survived the Library of Alexandria?',
      embedding: {
        provider: 'openai',
        model: 'text-embedding-3-large',
        dimensions: 1536,
        provider_key_ref: providerKeyRef,
      },
      inference: {
        provider: 'openai',
        model: 'gpt-4o-mini',
        provider_key_ref: providerKeyRef,
      },
    },
    { signal: ac.signal },
  );
  console.log(typeof r.answer === 'string' ? r.answer : JSON.stringify(r.answer, null, 2));
} catch (e) {
  if (e instanceof Error && e.name === 'AbortError') {
    console.error('cancelled');
    process.exit(2);
  }
  if (e instanceof TextralRetryExhausted) {
    const cause = e.cause instanceof TextralApiError ? `${e.cause.code}: ${e.cause.message}` : String(e.cause);
    console.error(`gave up after ${e.attempts} attempts — ${cause}`);
    process.exit(3);
  }
  throw e;
}

File: packages/sdk-python/examples/06_retry_and_cancel.py
"""06_retry_and_cancel.py — Retry policy + cancellation patterns.

Demonstrates: custom RetryPolicy with on_retry hook, NO_RETRY for
opt-out, asyncio cancellation propagating through the SDK.

Env: same as 01_quick_start.py.

Run:
    python 06_retry_and_cancel.py

See also:
    docs/development/sdks/SDK_COOKBOOK_OUTLINE.md §7
"""

from __future__ import annotations

import asyncio
import os

from textral import (
    NO_RETRY,
    AsyncClient,
    Client,
    RetryInfo,
    RetryPolicy,
    TextralAPIError,
    TextralRetryExhausted,
)

PROVIDER_KEY_REF = os.environ.get("TEXTRAL_PROVIDER_KEY_REF", "openai")
NAMESPACE = os.environ.get("TEXTRAL_NAMESPACE", "cookbook")


def demo_custom_retry() -> None:
    """Inject an `on_retry` hook so you can log every retry decision.
    Useful for diagnosing transient prod issues without enabling
    full debug logging."""

    def log_retry(info: RetryInfo) -> None:
        print(
            f"[retry] attempt={info.attempt} "
            f"status={info.status} "
            f"delay_ms={info.delay_ms} "
            f"err={type(info.error).__name__}"
        )

    policy = RetryPolicy(
        max_attempts=4,
        initial_delay_ms=300,
        max_delay_ms=5_000,
        on_retry=log_retry,
    )

    with Client(retry=policy) as client:
        try:
            client.me()
        except TextralRetryExhausted as e:
            print(f"retries exhausted after {e.attempts}: {e.last_error}")
        except TextralAPIError as e:
            print(f"non-retryable api error: {e}")


def demo_no_retry() -> None:
    """Disable retry entirely (e.g. for a one-shot interactive
    REPL). Errors surface on the first attempt."""
    with Client(retry=NO_RETRY) as client:
        try:
            client.me()
        except TextralAPIError as e:
            print(f"first-shot error: {e}")


async def demo_cancellation() -> None:
    """Cancel a long-running streaming query mid-flight. The async
    cancellation propagates through httpx → the SDK's stream
    iterator releases the response."""
    async with AsyncClient() as client:

        async def reader() -> None:
            async for frame in client.query.stream(
                namespace=NAMESPACE,
                query="Tell me everything about ancient libraries",
                embedding={
                    "provider": "openai",
                    "model": "text-embedding-3-large",
                    "dimensions": 1536,
                    "provider_key_ref": PROVIDER_KEY_REF,
                },
                inference={
                    "provider": "openai",
                    "model": "gpt-4o-mini",
                    "provider_key_ref": PROVIDER_KEY_REF,
                },
            ):
                if frame.get("event") == "answer_token":
                    print(frame.get("value", ""), end="", flush=True)

        task = asyncio.create_task(reader())
        await asyncio.sleep(2.0)  # Let some tokens flow.
        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            print("\n[cancelled]")


def main() -> None:
    print("=== custom retry ===")
    demo_custom_retry()
    print("\n=== no retry ===")
    demo_no_retry()
    print("\n=== cancellation ===")
    asyncio.run(demo_cancellation())


if __name__ == "__main__":
    main()
